### PR TITLE
Feature/exchanger autotests

### PR DESCRIPTION
**Pull request by scientistnik**

### CODE REVIEW

File: migrations/2_deploy_exchanger.js [hunk 1]
```diff
@@ -21,27 +21,55 @@ module.exports = async function(deployer, network) {
...
+        oraclePromise = oracles.map( (oracle) => {return oracle.deployed();});
+        oraclePromise.push(exchanger.deployed());
+        return Promise.all(oraclePromise);
+    })
+    .then( (contracts) => {
+        exch = contracts.pop();
+        return Promise.all(contracts.map( (oracle) => {return oracle.setBank(exch.address);}));})
+    .then( () => console.log("END DEPLOY"));
+  
+  //await applyDeps(contractsToDeploy,deployer,exchangerConfig);
```
Comment (decodedbrain) [inline migrations/2_deploy_exchanger.js]: Delete comments and unused methods

File: package.json [hunk 2]
```diff
@@ -19,5 +19,8 @@
   "bugs": {
     "url": "https://github.com/LibreCash/contracts/issues"
   },
-  "homepage": "https://github.com/LibreCash/contracts#readme"
+  "homepage": "https://github.com/LibreCash/contracts#readme",
+  "dependencies": {
+    "ganache-cli": "^6.1.0-beta.0"
```
Comment (decodedbrain) [inline package.json]: add truffle to deps

File: test/helpers/assertTx.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+function AssertTx() {
+
+    this.run = async (_func, _args) => {
+        var funcRes;
+        try {
+            funcRes = await _func(..._args);
+        } catch(e) {
+            if (e.toString().indexOf("VM Exception while processing transaction: revert") != -1) {
```
Comment (decodedbrain) [inline test/helpers/assertTx.js]: Replace to `~e.toString().indexOf("VM Exception while processing transaction: revert")`

File: test/testComplexExchanger.js [hunk 4]
```diff
@@ -0,0 +1,780 @@
+const
+    Reverter = require('./helpers/reverter'),
+    reverter = new Reverter(web3),
```
Comment (decodedbrain) [inline test/testComplexExchanger.js]: maybe use `const someVar = new (require(....))(params) ` if only one instance nedeed

File: test/testComplexExchanger.js [hunk 5]
```diff
@@ -0,0 +1,780 @@
...
+    "OracleMockKlara",
+    "OracleMockTest",
+    //"OracleMockRandom"
+].forEach( (filename) => {
+    oracles.push(artifacts.require(filename));
+});
+
+const delay = ms => new Promise(resolve => setTimeout(resolve, ms));
+
+function getWeiUsedForGas() {
```
Comment (decodedbrain) [inline test/testComplexExchanger.js]: maybe rename to gasCost

File: test/testComplexExchanger.js [hunk 6]
```diff
@@ -0,0 +1,780 @@
...
+                assert.equal((_oracleName.length - "0x".length) / 2, 32, "oracle name must be bytes32");
+                assert.equal((_oracleType.length - "0x".length) / 2, 16, "oracle type must be bytes16");
+                assert.equal(_waitQuery, false, "the new oracle shouldn't be waiting");
+                assert.equal(_updateTime.toNumber(), 0, "the new oracle's update time should be 0");
+                assert.equal(_callbackTime.toNumber(), 0, "the new oracle's update time should be 0");
+                assert.equal(_rate.toNumber(), 0, "the new oracle's rate should be 0");
+            }
+        });
+
+        /*
```
Comment (decodedbrain) [inline test/testComplexExchanger.js]: Uncomment or DELETE

File: test/testComplexExchanger.js [hunk 7]
```diff
@@ -0,0 +1,780 @@
...
+        it("(2) check PROCESSING_ORDERS", async function() {
+            let 
+                exchanger = await ComplexExchanger.deployed(),
+                state = + await exchanger.getState.call(),
+                calcTime = + await exchanger.calcTime.call();
+
+            if (now() - calcTime > exConfig.RATE_PERIOD) {
+                assert.notEqual(state, StateENUM.PROCESSING_ORDERS, 
+                    "Don't correct state when calcTime > RATE_PERIOD");
+                await exchanger.requestRates({value: web3.toWei(5,'ether')});
```
Comment (decodedbrain) [inline test/testComplexExchanger.js]: Move `5` to constants

File: test/testComplexExchanger.js [hunk 8]
```diff
@@ -0,0 +1,780 @@
...
+                await exchanger.calcRates();
+                state = + await exchanger.getState.call();
+            }
+
+            assert.equal(state, StateENUM.PROCESSING_ORDERS,"Don't correct state when calcTime < RATE_PERIOD");
+        });
+
+        it("(3) check WAIT_ORACLES", async function() {
+            let exchanger = await ComplexExchanger.deployed();
```
Comment (decodedbrain) [inline test/testComplexExchanger.js]: does we need to check it on every test? check docs pls

File: test/testComplexExchanger.js [hunk 9]
```diff
@@ -0,0 +1,780 @@
...
+
+            await exchanger.requestRates({value: web3.toWei(5,'ether')});
```
Comment (decodedbrain) [inline test/testComplexExchanger.js]: Move to constancts

File: test/testComplexExchanger.js [hunk 10]
```diff
@@ -0,0 +1,780 @@
...
+            let 
+                state = + await exchanger.getState.call(),
+                waiting = + await exchanger.waitingOracles.call();
+
+            if (waiting == 0) {
+                assert.notEqual(state, StateENUM.WAIT_ORACLES, "Don't correct state when waitingOracles == 0");
+                let oracle = await oracles[0].deployed();
```
Comment (decodedbrain) [inline test/testComplexExchanger.js]: Checke that we needs only one oracle

File: test/testComplexExchanger.js [hunk 11]
```diff
@@ -0,0 +1,780 @@
...
+            await exchanger.requestRates({value: web3.toWei(5,'ether')});
+
+            let 
+                state = + await exchanger.getState.call(),
+                ready = + await exchanger.readyOracles.call();
+
+            if (ready >= exConfig.MIN_READY_ORACLES) {
+                assert.equal(state, StateENUM.CALC_RATES, 
+                    "Don't correct state when readyOracles >= MIN_READY_ORACLES");
+                for(let i=0; i < (oracles.length - exConfig.MIN_READY_ORACLES + 1); i++) {
```
Comment (decodedbrain) [inline test/testComplexExchanger.js]: Check cycle condiftion. 
If we have 6 oracles, MIN_READY_ORACKES is 2:
6 - 2 + 1 = 5. 
Five is greater than 2, why we checks the state

File: test/testComplexExchanger.js (hunk 9 above)
Comment (scientistnik) [reply test/testComplexExchanger.js]: 5 it's random number, big then cost request.....why need create constant for it?

Comment (decodedbrain) [reply test/testComplexExchanger.js]: It's look like `magic-number` and if we need increase oracles count or use more expensive oracles we need to increase it. Const is great solution for this case.

File: test/testComplexExchanger.js (hunk 11 above)
Comment (scientistnik) [reply test/testComplexExchanger.js]: we check **state == CALC_RATES**, after that we broken 5 oracles, and check the **state != CALC_RATES**

File: test/testComplexExchanger.js (hunk 8 above)
Comment (scientistnik) [reply test/testComplexExchanger.js]: ne ponal))

**Pull request merged**